### PR TITLE
storage: prevent unbounded raft log growth without quorum

### LINKED ISSUE #27772: storage: excessively large raft logs
We've seen a handful of occurrences of excessively large Raft logs in clusters in the wild without being able to identify the root cause. A large Raft log should normally not occur as the proposal quota mechanism should limit the size of the Raft log if a range is healthy. If one node in the range is down the proposal quota mechanism will drop that replica from consideration, but at the same time allow Raft log truncation to occur, so the size of the Raft log should be limited to whatever Raft log truncation dictates (default 4MiB).

While each follower in a range maintains a Raft log, control of what gets written to the Raft log is in the hands of the leader. No leader and there are no Raft log writes. What happens if there is a leader but the Raft log entries it proposes are never getting applied? This can happen when the range is below quorum. Prior to 2.0, the Raft `CheckQuorum` mechanism would kick in and the leader would quickly step down. In 2.0 we turned on `PreVote` and disabled `CheckQuorum` (@bdarnell are they incompatible?). With `CheckQuorum` disabled and `PreVote` enabled a leader can remain the leader forever when the range is below quorum. The `PreVote` mechanism prevents another range from calling an election and without `CheckQuorum` enabled the leader won't step down. 

Note that the proposal quota mechanism still applies in this scenario and incoming operations will eventually block waiting for quota. There is another mechanism at work. The periodic Raft proposal refresh (which is necessary to deal with dropped proposals) interacts badly with this scenario. In particular, we refresh pending proposals every leader election timeout period (3s?). Refreshing a proposal only results in reproposing if the lease index is still compatible, but if a range is below quorum no other commands will be applied so it seems like we would see reproposals.

The above is a theory based on a reading of the code. It matches the conditions in the clusters that have experienced problems with large Raft logs (i.e. ranges that have gone through long periods of unavailability). @nvanbenschoten is going to work on writing a test to reproduce which should be straightforward to do if the above is correct. 

Fixing this problem shouldn't be too difficult. @bdarnell says:

> we want to stop refreshRaftProposals(reasonTicks) if liveness indicates that we don't have a quorum of live followers

and

> i think we might also want to only do reasonTicks refreshes if we are a follower
>
> the reason that path exists is due to raft's unacknowledged proposal forwarding
>
> if we handled ErrProposalDropped (https://github.com/cockroachdb/cockroach/issues/21849) and did our own forwarding, i think we might be able to get rid of reasonTicks refreshes completely

### PR DESCRIPTION
**Pull request by nvb**
Fixes #27772.

This change adds safeguards to prevent cases where a raft log
would grow without bound during loss of quorum scenarios. It
also adds a new test that demonstrates that the raft log does
not grow without bound in these cases.

There are two cases that need to be handled to prevent the
unbounded raft log growth observed in #27772.
1. When the leader proposes a command and cannot establish a
   quorum. In this case, we know the leader has the entry in
   its log, so there's no need to refresh it with `reasonTicks`.
   To avoid this, we no longer use `refreshTicks` as a leader.
2. When a follower proposes a command that is forwarded to the
   leader who cannot establish a quorum. In this case, the
   follower can't be sure (currently) that the leader got the
   proposal, so it needs to refresh using `reasonTicks`. However,
   the leader now detects duplicate forwarded proposals and
   avoids appending redundant entries to its log. It does so
   by maintaining a set of in-flight forwarded proposals that
   it has received during its term as leader. This set is reset
   after every leadership change.

Both of these cases are tested against in the new
TestLogGrowthWhenRefreshingPendingCommands. Without both of
the safeguards introduced in this commit, the test fails.

Release note (bug fix): Prevent loss of quorum situations from
allowing unbounded growth of a Range's Raft log.

### CODE REVIEW

**Comment by nvanbenschoten** [comment]
Extra logging in `refreshProposalsLocked` also confirmed that proposals were being refreshed with `reasonTicks` indefinitely. This isn't a surprise though.

**Review by petermattis — COMMENTED** [review]
Nice. Now to figure out a good fix. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27774#-:-LHp7GFZAKwKBbDTOk_S:b-23uvau)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Review by benesch — APPROVED** [review]
Well gosh. That sure is a simple test.

> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27774#-:-LHpjPPB4GTnjM9aGyTt:bmrm9px)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/client_raft_test.go, line 1168 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27774#-LHpjY7oCxDokvNpnN9d:-LHpjY7oCxDokvNpnN9e:bhsz42h) ([raw file](https://github.com/cockroachdb/cockroach/blob/184f08f0fb662de41bfabb7f260fcf8cf9694abf/pkg/storage/client_raft_test.go#L1168)):*
> ```Go
> 	putRes := make(chan *roachpb.Error)
> 	go func() {
> 		putArgs := putArgs([]byte("b"), make([]byte, 8<<10))
> ```

```
/* 8KiB */
```

**Review by petermattis — CHANGES_REQUESTED** [review]
Let's get either your solution in #27772 or Ben's proposed solution (don't refresh proposals on the leader for `reasonTicks`) into this PR. I have a preference for Ben's approach because we don't have to check `raft.Status` on every proposal. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27774#-:-LHryo_NB30F67Rlb6gd:b5xnqbl)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/client_raft_test.go, line 1174 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27774#-LHryYBbBkuxHdfQVOSb:-LHryYBbBkuxHdfQVOSc:b-qm4toi) ([raw file](https://github.com/cockroachdb/cockroach/blob/184f08f0fb662de41bfabb7f260fcf8cf9694abf/pkg/storage/client_raft_test.go#L1174)):*
> ```Go
> 
> 	// Wait for a bit and watch for Raft log growth.
> 	wait := time.After(30 * time.Second)
> ```

This makes the growth obvious, but does the test need to run this long to ensure that growth is not happening? I would think with a small enough raft tick interval we can run the test for a second or two and see unexpected Raft log growth.

**Comment by benesch** [comment]
Yeah, I didn't mean to mark this PR as "approved" but apparently Reviewable has a mind of its own these days.

**Comment by nvanbenschoten** [comment]
> Is that guaranteed to work fine? We could have put a proposal in when we weren't (Raft) leader, but then we get leadership and will never repropose them even though they're not going to pop out of the log.

I think the idea is that when getting leadership we would repropose the commands with a `reasonNewLeader` (or `reasonNewLeaderOrConfigChange`, I'm having trouble reasoning through when one applies over the other) and at that point the command will never need to be re-proposed unless we lose leadership again.

> If you're the leader the moment you propose there could still be a leadership loss and regain but that log index is lost.

Yes, that's a good point and a serious issue with the approach I proposed in https://github.com/cockroachdb/cockroach/issues/27772#issuecomment-406460761.

> Also, why don't we have that problem when commands are stuck on the follower? Wouldn't they still send a forwarded proposal to the leader every time, where it would get stuck in the log? This would be rare with 3x replication (if you have a follower you're committing anyway) but could happen with higher replication factors.

Agreed, this would all still break down if we were a follower in 5x replication group where all other followers went down but the leader stayed up. We would just keep forwarding the proposal to the leader and keep appending to its log.

#27054 does sound like it would help improve the general situation with `reasonTicks`, but I'm worried about relying on these proposal retries any more than we already do. One reason for this is that proposals that were pushed through Raft asynchronously will not retry and will instead result in an entire txn retry.

@tschottdorf I'm curious what your take on @bdarnell's [mention](https://github.com/cockroachdb/cockroach/issues/27772#issue-342926152) of #21849 is. Finding all the scenerios where `reasonTicks` reproposals can enter a hot loop seems fraught to me, so the idea of eliminating it entirely sounds very appealing.

**Comment by bdarnell** [comment]
> but could happen with higher replication factors.

You're right. Doing anything based on leadership status only works for 3x replication. 

And my other proposal, to skip reasonTicks when we don't have a quorum of known-live nodes, is likely subject to problems when a reproposal is needed to unblock the liveness range itself.

Handling ErrProposalDropped and taking control of proposal forwarding definitely looks like the right long-term fix, but we couldn't backport it to 2.0. 

How about this: Leaders skip reasonTicks reproposals, *and* drop incoming MsgProps if the uncommitted portion of the raft log is too large.

**Comment by nvanbenschoten** [comment]
> and drop incoming MsgProps if the uncommitted portion of the raft log is too large.

I'm a little worried about this too. What happens in the case where the leaseholder is not the leader and we start dropping all forwarded MsgProps? How does the raft log ever get truncated in that situation? Do we just wait for the leaseholder and raft leader to collocate before making any more progress?

I think I'm coming around to the idea of looking at node liveness in 2.0 and fixing this properly in 2.1. We could ignore the node liveness check on the node liveness range itself if we are afraid of creating a dependency cycle between node liveness proposals and the node liveness state itself.

**Comment by bdarnell** [comment]
> I'm a little worried about this too. What happens in the case where the leaseholder is not the leader and we start dropping all forwarded MsgProps?

The leader will continue to send MsgApps to followers when and if they are healthy, eventually committing the outstanding proposals and making room for new ones. This all happens at the raft level independently of the lease. A lease will be needed to truncate the raft log, but this should proceed normally once quorum is reestablished.

**Comment by nvanbenschoten** [comment]
Ok, this is ready for a review. I went a slightly different route with the approach to preventing duplicate forwarded proposals (see PR description), but it's actually pretty similar to @bdarnell's final comment. Completely avoiding duplicate forwarded proposals by tracking them on the leader makes this easier to test and feels a little more safe to me. I don't see any real issues with the other approach though.

**Review by petermattis — DISMISSED** [review]
I have to take another look at this tomorrow. One quick comment for tonight.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27774#-:-LI3N1sSCGge3pUP0KwL:b-wg7ka2)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/replica.go, line 390 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27774#-LHxGfLa1jSvzwjM6k2G:-LHxGfLb7e92c8vQNjX8:b1lpxky) ([raw file](https://github.com/cockroachdb/cockroach/blob/d23a57c31292919f1ee8843577c0b99c5a642d00/pkg/storage/replica.go#L390)):*
> ```Go
> 		// The set allows leaders to detect duplicate forwarded commands and
> 		// avoid re-proposing the same forwarded command multiple times.
> 		proposalsForwarded map[storagebase.CmdIDKey]struct{}
> ```

`proposalsForwarded` sounds like it would be maintained on the follower. I would have named this `receivedProposals` or `remoteProposals`. We could also rename the `proposals` field to `localProposals`.

**Review by bdarnell — APPROVED** [review]
:lgtm: 

Release note: This prevents out-of-memory crashes, not "loss of quorum situations". 

> Reviewed 4 of 6 files at r3, 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27774#-:-LI4DYh4-J2l33VWc6PV:bu7pmg1)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/replica.go, line 3776 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/27774#-LI4D09D201KpD6kFFg0:-LI4D09D201KpD6kFFg1:b-krrj26) ([raw file](https://github.com/cockroachdb/cockroach/blob/7306d9f99ef778b5251056c4504a4e279d70f8c4/pkg/storage/replica.go#L3776)):*
> ```Go
> 						}
> 					case raftpb.EntryConfChange:
> 						allSeen = false
> ```

Comment that we could peek into the EntryConfChange to find the command ID, but we don't expect follower-initiated conf changes. 
___
*[pkg/storage/replica.go, line 3896 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/27774#-LI4CcDRDxjFk9_MiklS:-LI4CcDRDxjFk9_MiklT:b-g11lgo) ([raw file](https://github.com/cockroachdb/cockroach/blob/7306d9f99ef778b5251056c4504a4e279d70f8c4/pkg/storage/replica.go#L3896)):*
> ```Go
> 		// Clear the forwarded proposal set. No-op if not previously the leader.
> 		for cmdID := range r.mu.proposalsForwarded {
> 			delete(r.mu.proposalsForwarded, cmdID)
> ```

Reset the map to nil instead of deleting all the entries. We're unlikely to need the map again soon. 
___
*[pkg/storage/replica_test.go, line 8132 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/27774#-LI4BfHy5HLTVaAE-y3L:-LI4BfHy5HLTVaAE-y3M:b1dlthq) ([raw file](https://github.com/cockroachdb/cockroach/blob/7306d9f99ef778b5251056c4504a4e279d70f8c4/pkg/storage/replica_test.go#L8132)):*
> ```Go
> 	// Enable so that pending commands are reproposed even though the replica
> 	// will become the Raft leader.
> 	cfg.RefreshPendingCommandsWhenLeader = true
> ```

I don't think it makes sense to add a new option so that this test can go on working like it was. This test is no longer verifying the behavior of reasonTicks refreshes as seen in production, so it should either be updated to match the new behavior (that is, by making its proposals through a follower so the reasonTicks reproposal is necessary) or be deleted.

**Review by nvb — COMMENTED** [review]
> Release note: This prevents out-of-memory crashes, not "loss of quorum situations".

Hah, I wish. I think you missed the next line of the release notes. "Prevent loss of quorum situations *from allowing unbounded growth of a Range's Raft log.*"

> Handling ErrProposalDropped and taking control of proposal forwarding definitely looks like the right long-term fix, but we couldn't backport it to 2.0.

I'm actually not so sure about this anymore. Never refreshing with `reasonTicks` on the leader is a solid improvement on its own and tracking successfully forwarded proposals gets us the other half of the way there. Correct me if I'm wrong, but I believe the primary difference between "handling ErrProposalDropped" and what we do here is that we're tracking whether forwarded proposals succesfully made it into the Raft log on the leader here instead of on the proposer/follower like we do in the other approach. In the other approach, we would still need to re-propose and re-forward whenever the leader changed so we would still need similar bookkeeping. I guess the main benefit of the other approach is avoiding the continuous forwarding (that now becomes a no-op) under a steady leader when nothing is being committed.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27774#-:-LI6I4fYB-L4-GTR257v:b-4xkcy7)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/client_raft_test.go, line 1168 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27774#-LHpjY7oCxDokvNpnN9d:-LI6IJrN48AoXh348tyc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/184f08f0fb662de41bfabb7f260fcf8cf9694abf/pkg/storage/client_raft_test.go#L1168)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

```
/* 8KiB */
```
</blockquote></details>

Done.
___
*[pkg/storage/client_raft_test.go, line 1174 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27774#-LHryYBbBkuxHdfQVOSb:-LI6IKZ4AhSh4Yy8DDzS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/184f08f0fb662de41bfabb7f260fcf8cf9694abf/pkg/storage/client_raft_test.go#L1174)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This makes the growth obvious, but does the test need to run this long to ensure that growth is not happening? I would think with a small enough raft tick interval we can run the test for a second or two and see unexpected Raft log growth.
</blockquote></details>

Done.
___
*[pkg/storage/replica.go, line 390 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27774#-LHxGfLa1jSvzwjM6k2G:-LI6IOubBz-abgazo3i0:b3blvmt) ([raw file](https://github.com/cockroachdb/cockroach/blob/d23a57c31292919f1ee8843577c0b99c5a642d00/pkg/storage/replica.go#L390)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`proposalsForwarded` sounds like it would be maintained on the follower. I would have named this `receivedProposals` or `remoteProposals`. We could also rename the `proposals` field to `localProposals`. 
</blockquote></details>

Done. Went with `remoteProposals` and renamed `proposals` to `localProposals`.
___
*[pkg/storage/replica.go, line 3776 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/27774#-LI4D09D201KpD6kFFg0:-LI6L-7Z0cURBsSUPeMH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7306d9f99ef778b5251056c4504a4e279d70f8c4/pkg/storage/replica.go#L3776)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Comment that we could peek into the EntryConfChange to find the command ID, but we don't expect follower-initiated conf changes. 
</blockquote></details>

Done.
___
*[pkg/storage/replica.go, line 3896 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/27774#-LI4CcDRDxjFk9_MiklS:-LI6MLqX9NqwJ84nnoC7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7306d9f99ef778b5251056c4504a4e279d70f8c4/pkg/storage/replica.go#L3896)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Reset the map to nil instead of deleting all the entries. We're unlikely to need the map again soon. 
</blockquote></details>

Done.
___
*[pkg/storage/replica_test.go, line 8132 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/27774#-LI4BfHy5HLTVaAE-y3L:-LI6kuZdBJ16NLzy31HG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7306d9f99ef778b5251056c4504a4e279d70f8c4/pkg/storage/replica_test.go#L8132)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I don't think it makes sense to add a new option so that this test can go on working like it was. This test is no longer verifying the behavior of reasonTicks refreshes as seen in production, so it should either be updated to match the new behavior (that is, by making its proposals through a follower so the reasonTicks reproposal is necessary) or be deleted. 
</blockquote></details>

Done.

**Review by bdarnell — APPROVED** [review]
> Hah, I wish. I think you missed the next line of the release notes. "Prevent loss of quorum situations from allowing unbounded growth of a Range's Raft log."

I saw that, I just think that OOM crashes better captures the user-observable behavior here. It doesn't necessarily lead to loss of quorum (if the other nodes decide to go on without the former leader that is filling up its log, the quorum is fine).

> I guess the main benefit of the other approach is avoiding the continuous forwarding (that now becomes a no-op) under a steady leader when nothing is being committed.

Right. Changing the way proposals are forwarded would be a cleanup and allow us to get rid of `reasonTicks` completely, but it wouldn't have any advantage in terms of raft log growth over what you're doing here. 

> Reviewed 2 of 5 files at r5, 3 of 3 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27774#-:-LI81Lbf3pxOtmvg6Slg:bkqu2a9)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/client_raft_test.go, line 1151 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/27774#-LI83beZ8yCi46XfyhyH:-LI83beZ8yCi46XfyhyI:b-2nj0xx) ([raw file](https://github.com/cockroachdb/cockroach/blob/99615e1603d11e5268e23e1c9955b355bbb3eeee/pkg/storage/client_raft_test.go#L1151)):*
> ```Go
> 	// Refresh pending commands on every Raft group tick instead of
> 	// every RaftElectionTimeoutTicks.
> 	sc.TestingKnobs.RefreshReasonTicksPeriodicity = 3
> ```

Shouldn't this be 1 to match the comment? (or the comment should be updated)
___
*[pkg/storage/store.go, line 774 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/27774#-LI83C0CBM42Azgq6ozy:-LI83C0CBM42Azgq6ozz:b-boxoz1) ([raw file](https://github.com/cockroachdb/cockroach/blob/99615e1603d11e5268e23e1c9955b355bbb3eeee/pkg/storage/store.go#L774)):*
> ```Go
> 	// periodically.
> 	DisableRefreshReasonTicks bool
> 	// RefreshReasonTicksPeriodicity overrides the default periodicity in which
> ```

s/icity//g (or s/periodicity/interval/g)

**Review by nvb — COMMENTED** [review]
> I saw that, I just think that OOM crashes better captures the user-observable behavior here. It doesn't necessarily lead to loss of quorum (if the other nodes decide to go on without the former leader that is filling up its log, the quorum is fine).

I still don't think we're seeing eye to eye. I'm not saying that this was ever causing a loss of quorum, I'm saying that a loss of quorum would previously cause unbounded log growth and that this prevents this adverse effect. Maybe the release note is confusingly worded?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27774#-:-LI86KBD9GFZ_fwisBOw:b-h1pzut)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/client_raft_test.go, line 1151 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/27774#-LI83beZ8yCi46XfyhyH:-LI85oHjAVZC1cii4g-t:b-of6bo5) ([raw file](https://github.com/cockroachdb/cockroach/blob/99615e1603d11e5268e23e1c9955b355bbb3eeee/pkg/storage/client_raft_test.go#L1151)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Shouldn't this be 1 to match the comment? (or the comment should be updated)
</blockquote></details>

Done. Good catch, I was playing with this to make sure it was working as expected and forgot to bring it back down to 1.
___
*[pkg/storage/store.go, line 774 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/27774#-LI83C0CBM42Azgq6ozy:-LI8650HC47pMD9Euc1v:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99615e1603d11e5268e23e1c9955b355bbb3eeee/pkg/storage/store.go#L774)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/icity//g (or s/periodicity/interval/g)
</blockquote></details>

Done.

**Review by bdarnell — COMMENTED** [review]
Ah, I see. I parse that sentence as "(Prevent loss of quorum situations) from [caused by] allowing unbounded growth of a Range's Raft log"  but you meant "Prevent (loss of quorum situations from allowing unbounded growth of a Range's Raft lo)". I think it would be clearer as "Prevent unbounded growth of the raft log caused by a loss of quorum". 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27774#-:-LI87bdOCr1gpYo-ytKK:b-ruv3co)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Comment by nvanbenschoten** [comment]
Done. Glad neither of us was going crazy.

bors r+

File: pkg/storage/client_raft_test.go [hunk 1]
```diff
@@ -1126,6 +1127,144 @@ func TestRefreshPendingCommands(t *testing.T) {
 	}
 }
 
+// Test that when a Raft group is not able to establish a quorum, its Raft log
+// does not grow without bound. It tests two different scenerios where this used
```
Comment (tschottdorf) [inline pkg/storage/client_raft_test.go]: scenerios 💩

Comment (nvanbenschoten) [reply pkg/storage/client_raft_test.go]: Way ahead of you https://github.com/cockroachdb/cockroach/commit/d41111da6740b3e07bdf9f00724606b455e6e11f#diff-b1bd51088f60beeacafe950f5bc700f8R1131 😄

**Pull request merged**